File: src/ts/src/local-file-scoped.ts
import {
  LocalFileLock,
  acquire_local_file_lock,
  type LocalFileLockOptions,
} from "./local-file.js";

export type ScopedLocalFileLockErrorKind = "lock" | "work" | "work_and_release";

/** Structured outcome preserving work and cleanup failures independently. */
export class ScopedLocalFileLockError extends Error {
  readonly kind: ScopedLocalFileLockErrorKind;
  readonly lock_error: unknown | undefined;
  readonly work_error: unknown | undefined;

  constructor(
    kind: ScopedLocalFileLockErrorKind,
    message: string,
    options: { lock_error?: unknown; work_error?: unknown } = {},
  ) {
    super(message);
    this.name = "ScopedLocalFileLockError";
    this.kind = kind;
    this.lock_error = options.lock_error;
    this.work_error = options.work_error;
  }
}

/**
 * Acquire, run one callback, and release exactly once.
 *
 * Expected callback failures may be any thrown value. Fatal process aborts are
 * outside this structured contract. When work and release both fail, neither
 * failure is discarded.
 */
export async function with_local_file_lock<T>(
  path: string,
  owner: string,
  options: LocalFileLockOptions,
  work: (lock: LocalFileLock) => T | Promise<T>,
): Promise<T> {
  let lock: LocalFileLock;
  try {
    lock = await acquire_local_file_lock(path, owner, options);
  } catch (lock_error) {
    throw new ScopedLocalFileLockError("lock", "local lock acquisition failed", { lock_error });
  }

  let value: T | undefined;
  let work_error: unknown | undefined;
  try {
    value = await work(lock);
  } catch (error) {
    work_error = error;
  }

  let lock_error: unknown | undefined;
  try {
    await lock.release();
  } catch (error) {
    lock_error = error;
  }

  if (work_error !== undefined && lock_error !== undefined) {
    throw new ScopedLocalFileLockError(
      "work_and_release",
      "local lock work and release both failed",
      { work_error, lock_error },
    );
  }
  if (work_error !== undefined) {
    throw new ScopedLocalFileLockError("work", "local lock work failed", { work_error });
  }
  if (lock_error !== undefined) {
    throw new ScopedLocalFileLockError("lock", "local lock release failed", { lock_error });
  }

  return value as T;
}

File: src/go/local_file_scoped.go
package oreslocks

// ScopedLocalFileLockErrorKind classifies structured failures from
// WithLocalFileLock without discarding either work or release errors.
type ScopedLocalFileLockErrorKind string

const (
	ScopedLocalFileLockErrorLock           ScopedLocalFileLockErrorKind = "lock"
	ScopedLocalFileLockErrorWork           ScopedLocalFileLockErrorKind = "work"
	ScopedLocalFileLockErrorWorkAndRelease ScopedLocalFileLockErrorKind = "work_and_release"
)

// ScopedLocalFileLockError preserves the work and/or local-lock failure that
// produced a scoped-operation error.
type ScopedLocalFileLockError struct {
	Kind      ScopedLocalFileLockErrorKind
	LockError error
	WorkError error
}

func (e *ScopedLocalFileLockError) Error() string {
	switch e.Kind {
	case ScopedLocalFileLockErrorWorkAndRelease:
		return "local lock work and release both failed"
	case ScopedLocalFileLockErrorWork:
		return "local lock work failed"
	default:
		return "local lock acquisition or release failed"
	}
}

// WithLocalFileLock acquires, executes work exactly once, then releases exactly
// once after a successful acquisition. Expected work failures are returned as
// error values. Go panics are not normalized into structured work errors; defer
// attempts release during unwinding and the original panic keeps precedence.
func WithLocalFileLock[T any](
	path string,
	owner string,
	options LocalFileLockOptions,
	work func(*LocalFileLock) (T, error),
) (value T, err error) {
	lock, acquireErr := AcquireLocalFileLock(path, owner, options)
	if acquireErr != nil {
		return value, &ScopedLocalFileLockError{
			Kind:      ScopedLocalFileLockErrorLock,
			LockError: acquireErr,
		}
	}

	// Preserve fatal panic precedence while still making a best-effort release.
	deferredRelease := true
	defer func() {
		if deferredRelease {
			_ = lock.Release()
		}
	}()

	value, workErr := work(lock)
	releaseErr := lock.Release()
	deferredRelease = false

	switch {
	case workErr != nil && releaseErr != nil:
		return value, &ScopedLocalFileLockError{
			Kind:      ScopedLocalFileLockErrorWorkAndRelease,
			WorkError: workErr,
			LockError: releaseErr,
		}
	case workErr != nil:
		return value, &ScopedLocalFileLockError{
			Kind:      ScopedLocalFileLockErrorWork,
			WorkError: workErr,
		}
	case releaseErr != nil:
		return value, &ScopedLocalFileLockError{
			Kind:      ScopedLocalFileLockErrorLock,
			LockError: releaseErr,
		}
	default:
		return value, nil
	}
}
